db/util_testing.go

```diff
@@ -172,11 +177,78 @@ func WaitForUserWaiterChange(userWaiter *ChangeWaiter) bool {
 	return isChanged
 }
 
+// emptyAllDocsIndex ensures the AllDocs index for the given bucket is empty. Works similarly to db.Compact, except on a different index.
+func emptyAllDocsIndex(ctx context.Context, b *base.CouchbaseBucketGoCB, tbp *base.TestBucketPool) (numCompacted int, err error) {
+	purgedDocCount := 0
+	purgeBody := Body{"_purged": true}
+
+	dbCtx, err := NewDatabaseContext("db", base.NoCloseClone(b), false, DatabaseContextOptions{
+		UseViews:    base.TestsDisableGSI(),
+		EnableXattr: base.TestUseXattrs(),
+	})
+	if err != nil {
+		return 0, err
+	}
+	defer dbCtx.Close()
+
+	database, err := GetDatabase(dbCtx, nil)
+	if err != nil {
+		return 0, err
+	}
+
+	// A stripped down version of db.Compact() that works on AllDocs instead of tombstones
+	for {
+		purgedDocs := make([]string, 0)
+		results, err := database.QueryChannels("*", 0, 0, 0, false)
+		if err != nil {
+			return 0, err
+		}
+		var tombstonesRow QueryIdRow
+		var resultCount int
+		for results.Next(&tombstonesRow) {
+			resultCount++
+			tbp.Logf(ctx, "compactTestBucket deleting %q", tombstonesRow.Id)
+			// First, attempt to purge.
+			purgeErr := database.Purge(tombstonesRow.Id)
+			if purgeErr == nil {
+				purgedDocs = append(purgedDocs, tombstonesRow.Id)
+			} else if base.IsKeyNotFoundError(b, purgeErr) {
+				// If key no longer exists, need to add and remove to trigger removal from view
+				_, addErr := b.Add(tombstonesRow.Id, 0, purgeBody)
+				if addErr != nil {
+					tbp.Logf(ctx, "Error compacting key %s (add) - will not be compacted.  %v", tombstonesRow.Id, addErr)
+					continue
+				}
+
+				// At this point, the doc is not in a usable state for mobile
+				// so mark it to be removed from cache, even if the subsequent delete fails
+				purgedDocs = append(purgedDocs, tombstonesRow.Id)
+
+				if delErr := b.Delete(tombstonesRow.Id); delErr != nil {
+					tbp.Logf(ctx, "Error compacting key %s (delete) - will not be compacted.  %v", tombstonesRow.Id, delErr)
+				}
+			} else {
+				tbp.Logf(ctx, "Error compacting key %s (purge) - will not be compacted.  %v", tombstonesRow.Id, purgeErr)
+			}
+		}
+
+		purgedDocCount += len(purgedDocs)
```

Comment:
Looks like this is the only remaining usage for purgedDocs - could just replace with a counter at this point?

Reply:
Removed purgedDocs slice and just update purgedDocCount directly.
Had to rebase for conflicts, but no further changes.